fortio 0.11.0 (new formula)

ymesika opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Closed PR #26430 and re-created a new one due to rebase mistake I made. Sorry for the mess.
@fxcoudert I followed your change requests in the closed PR and fixed the formula.

CC @ldemailly

ldemailly commented on Formula/fortio.rb:
```diff
@@ -4,8 +4,8 @@ class Fortio < Formula
   sha256 "76871124e7abd13d7fa0ccc5ad6912bc1c3d77d53694da2b26ad44af3638edab"
 
-  depends_on "go" => :build
   depends_on "dep" => :build
```
dep shouldn’t be needed with submodule (but I don’t know if the submodule is in the tar ball)

ldemailly commented on Formula/fortio.rb:
```diff
@@ -0,0 +1,33 @@
+  end
+
+  test do
+    assert_match version.to_s, shell_output("#{bin}/fortio --version")
```
please use `fortio version -s`

woodruffw commented:
`test` failures on El Cap and Sierra:

```
==> /usr/local/Cellar/fortio/0.9.0/bin/fortio version -s
Error: fortio: failed
<true> expected but was
<false>.
```

ymesika commented:
@woodruffw Thanks for notifying us.
Tests were updated to avoid the failures.

ilovezfs commented on Formula/fortio.rb:
```diff
@@ -0,0 +1,39 @@
+  test do
+    assert_match version.to_s, shell_output("#{bin}/fortio version -s")
+    fortio_srv = fork do
+      exec "#{bin}/fortio server -http-port 8085"
```
please avoid hard-coding the port
```
    require "socket"

    server = TCPServer.new(0)
    port = server.addr[1]
    server.close
```
and use port

ymesika replied:
Nice suggestion to avoid port conflicts

ilovezfs commented on Formula/fortio.rb:
```diff
@@ -0,0 +1,39 @@
+    (buildpath/"src/istio.io/fortio").install contents
+    cd "src/istio.io/fortio" do
+      system "dep", "ensure"
+      date = `date +"%Y-%m-%d %H:%M"`.strip
```
Utils.popen_read instead of backticks

ymesika replied:
Replaced

ilovezfs commented on Formula/fortio.rb:
```diff
@@ -0,0 +1,39 @@
+      exec "#{bin}/fortio server -http-port 8085"
+    end
+    sleep 2
+    load_res = `#{bin}/fortio load -quiet http://localhost:8085/ 2>&1 | tail -1`
```
Utils.popen_read instead of backticks and stick to Ruby instead of using tail

ymesika replied:
Replaced

ilovezfs commented on Formula/fortio.rb:
```diff
@@ -0,0 +1,39 @@
+    sleep 2
+    load_res = `#{bin}/fortio load -quiet http://localhost:8085/ 2>&1 | tail -1`
+    Process.kill("SIGTERM", fortio_srv)
+    assert_equal true, load_res.start_with?("All done")
```
use assert_match with a regex starting with `^`

ymesika replied:
Replaced

ilovezfs commented on Formula/fortio.rb:
```diff
@@ -0,0 +1,39 @@
+    ENV["XC_ARCH"] = "amd64"
+    ENV["GOPATH"] = buildpath
+    ENV["CGO_ENABLED"] = "0"
+    contents = Dir["{*,.git,.gitignore}"]
```
what happens if you just use `buildpath.children`?

ymesika replied:
No issues with it

ilovezfs commented on Formula/fortio.rb:
```diff
@@ -0,0 +1,39 @@
+    ENV["XC_OS"] = "darwin"
+    ENV["XC_ARCH"] = "amd64"
+    ENV["GOPATH"] = buildpath
+    ENV["CGO_ENABLED"] = "0"
```
why?

ymesika replied:
Removed the CGO_ENABLED flag without any issues.
@ldemailly FYI

ldemailly replied:
CGO_ENABLED=0 is to get truly static binaries that do not depend on go installation; I would think it is necessary (or at least does not hurt, the fortio binary is small)

MikeMcQuaid replied:
> CGO_ENABLED=0 is to get truly static binaries that do not depend on go installation

This isn't the case for most software, why is it for Fortio?



> or at least does not hurt

We don't want to add things unless they are 100% necessary.

ldemailly replied:
thanks. can you point me to another go program tap as an example ?
(fortio ships as a static binary for docker and linux to be "copyable" I thought it's a useful feature for mac os as well, I am not sure how it even works dynamically on a host that doesn't have go runtime installed)

MikeMcQuaid replied:
@ldemailly https://github.com/Homebrew/homebrew-core/blob/master/Formula/fork-cleaner.rb. You shouldn't have to remove `CGO` support for it to not depend on `go` at runtime.

ilovezfs commented on Formula/fortio.rb:
```diff
@@ -0,0 +1,39 @@
+      link_flags = "-s -X istio.io/fortio/ui.resourcesDir=#{lib} "\
+        "-X istio.io/fortio/version.tag=v#{version} "\
+        "-X \"istio.io/fortio/version.buildInfo=#{date}\""
+      system "go", "build", "-a", "-ldflags", link_flags.to_s
```
use `-o` instead of `bin.install`

ymesika replied:
Replaced

ilovezfs commented on Formula/fortio.rb:
```diff
@@ -0,0 +1,39 @@
+    contents = Dir["{*,.git,.gitignore}"]
+    (buildpath/"src/istio.io/fortio").install contents
+    cd "src/istio.io/fortio" do
+      system "dep", "ensure"
```
If we use the Git tag to pull the submodules, is this needed? There's seems to be a vendor directory.

ymesika replied:
In a previous review we have been asked to use the source tarball rather than a git pull.
Unfortunately the tarball doesn't have the content of the `vendor` folder and therefore the use of `dep ensure` to fetch it.

ldemailly commented on Formula/fortio.rb:
```diff
@@ -11,29 +11,34 @@ def install
+    server.close
     fortio_srv = fork do
-      exec "#{bin}/fortio server -http-port 8085"
+      exec "#{bin}/fortio server -http-port #{port}"
```
should be
```
fortio server --quiet --redirect-port "disabled" --grpc-port "disabled" -http-port 0 -ui-path ""
```
and get the URL from stdout $7 . (`|awk '{print $7}'`)

ldemailly replied:
for instance
```
docker run istio/fortio:0.9.0 server --quiet --redirect-port "disabled" --grpc-port "disabled" -http-port 0 -ui-path "" 2> /dev/null | awk '{print $7}'
[::]:36593
```
I'm fixing the stderr spurious message for grpc disable btw, and will make that that --quiet only prints the url

ymesika replied:
Can we have this change in 0.9.1?

ldemailly replied:
yes sure if this passes as is it can wait

ymesika commented:
@ilovezfs Thanks for your feedback. We've updated the formula based on it.

ldemailly commented on Formula/fortio.rb:
```diff
@@ -0,0 +1,44 @@
+  def install
+    ENV["XC_OS"] = "darwin"
+    ENV["XC_ARCH"] = "amd64"
+    ENV["GOPATH"] = buildpath
```
I'd put back CGO_ENABLED=0 if you don't mind (or also we can do that in next rev)

ldemailly commented:
@ilovezfs ping :-) I'd love to have this in place for kubecon next week where we'll demo fortio more widely

ldemailly commented:
ping ping :)

ldemailly commented:
poke?

ldemailly commented:
@mikemcquaid sorry to nag but would you have a chance to help us get this in ?

Thanks!

MikeMcQuaid commented:
@ldemailly @ilovezfs will need to approve this rather than me.

ldemailly commented:
@MikeMcQuaid we made the changes @ilovezfs requested, anyone (else if @ilovezfs is busy/out - hope everything is ok) able to approve this ? it's been a month

MikeMcQuaid commented on Formula/fortio.rb:
```diff
@@ -0,0 +1,45 @@
+    cd "src/istio.io/fortio" do
+      system "dep", "ensure"
+      date = Utils.popen_read("date", "+%Y-%m-%d %H:%M").strip
+      link_flags = "-s -X istio.io/fortio/ui.resourcesDir=#{lib} "\
```
No reason for this to be be a variable and each e.g. `-s`, `-X` should be a separate argument.

MikeMcQuaid commented:
@ldemailly You didn't really address the `CGO_ENABLED` comment. As a more general comment in addition to mine above: this is really crying out for a `Makefile` that sets the necessary environment variables, link flags and provides a `make install` target. It's unclear how anyone would know what these variables/link flags/files to install should be and how they'll need to be updated for future versions.

ldemailly commented:
the release/build process, ignoring brew/macos, is in 1 place today:
https://github.com/istio/fortio/blob/master/Dockerfile#L12-L17
but we can move some of that logic in the Makefile to be shared with macos build

(dev builds are just "go install" so only release need all the extra logic)

ymesika commented:
@MikeMcQuaid Thanks for your review. Updated:
1. Removed `CGO_ENABLED` flag
1. Got rid of the `ldflags` variable. Can't break the flag value into separated strings as its all one argument value. Tried that and made sure that other formulas are indeed doing the same
1. Run install, test and strict audit
1. Updated to latest _fortio_ version (0.11.0)

MikeMcQuaid commented on Formula/fortio.rb:
```diff
@@ -0,0 +1,44 @@
+    sleep 2
+
+    load_res = Utils.popen_read("#{bin}/fortio", "load", "http://localhost:#{port}/", :err => :out)
+    Process.kill("SIGTERM", fortio_srv)
```
Probably want the `.kill` in an `ensure` to make sure it will always be killed.

MikeMcQuaid commented on Formula/fortio.rb:
```diff
@@ -0,0 +1,44 @@
+    (buildpath/"src/istio.io/fortio").install buildpath.children
+    cd "src/istio.io/fortio" do
+      system "dep", "ensure"
+      date = Utils.popen_read("date", "+%Y-%m-%d %H:%M").strip
```
e.g. `DateTime` or `Time`.

ymesika commented:
@MikeMcQuaid Thanks. Comments addressed.

ymesika commented:
@ilovezfs @MikeMcQuaid Thanks!

ilovezfs commented:
@ymesika You're welcome!
